File: lib/java/src/main/java/uapi/internal/schema/CatchHeaderCollisions.java
package uapi.internal.schema;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import uapi.UApiSchemaParseError;

import static uapi.internal.schema.GetPathDocumentCoordinatesPseudoJson.getPathDocumentCoordinatesPseudoJson;

public class CatchHeaderCollisions {

    public static void catchHeaderCollisions(
            Map<String, List<Object>> uApiSchemaNameToPseudoJson,
            Set<String> headerKeys,
            Map<String, Integer> keysToIndex,
            Map<String, String> schemaKeysToDocumentNames,
            Map<String, String> documentNamesToJson) {
        List<SchemaParseFailure> parseFailures = new ArrayList<>();

        List<String> headerKeysList = new ArrayList<>(headerKeys);

        headerKeysList.sort((k1, k2) -> {
            String documentName1 = schemaKeysToDocumentNames.get(k1);
            String documentName2 = schemaKeysToDocumentNames.get(k2);
            if (!documentName1.equals(documentName2)) {
                return documentName1.compareTo(documentName2);
            } else {
                int index1 = keysToIndex.get(k1);
                int index2 = keysToIndex.get(k2);
                return Integer.compare(index1, index2);
            }
        });

        for (int i = 0; i < headerKeysList.size(); i++) {
            for (int j = i + 1; j < headerKeysList.size(); j++) {
                String defKey = headerKeysList.get(i);
                String otherDefKey = headerKeysList.get(j);

                int index = keysToIndex.get(defKey);
                int otherIndex = keysToIndex.get(otherDefKey);

                String documentName = schemaKeysToDocumentNames.get(defKey);
                String otherDocumentName = schemaKeysToDocumentNames.get(otherDefKey);

                List<Object> uApiSchemaPseudoJson = uApiSchemaNameToPseudoJson.get(documentName);
                List<Object> otherUApiSchemaPseudoJson = uApiSchemaNameToPseudoJson.get(otherDocumentName);

                Map<String, Object> def = (Map<String, Object>) uApiSchemaPseudoJson.get(index);
                Map<String, Object> otherDef = (Map<String, Object>) otherUApiSchemaPseudoJson.get(otherIndex);

                Map<String, Object> headerDef = (Map<String, Object>) def.get(defKey);
                Map<String, Object> otherHeaderDef = (Map<String, Object>) otherDef.get(otherDefKey);

                Set<String> headerCollisions = new HashSet<>(headerDef.keySet());
                headerCollisions.retainAll(otherHeaderDef.keySet());
                for (String headerCollision : headerCollisions) {
                    List<Object> thisPath = Arrays.asList(index, defKey, headerCollision);
                    String thisDocumentJson = documentNamesToJson.get(documentName);
                    Object thisLocation = getPathDocumentCoordinatesPseudoJson(thisPath, thisDocumentJson);
                    parseFailures.add(new SchemaParseFailure(
                            otherDocumentName,
                            Arrays.asList(otherIndex, otherDefKey, headerCollision),
                            "PathCollision",
                            Map.of("document", documentName, "path", thisPath, "location", thisLocation)));
                }

                Map<String, Object> resHeaderDef = (Map<String, Object>) def.get("->");
                Map<String, Object> otherResHeaderDef = (Map<String, Object>) otherDef.get("->");

                Set<String> resHeaderCollisions = new HashSet<>(resHeaderDef.keySet());
                resHeaderCollisions.retainAll(otherResHeaderDef.keySet());
                for (String resHeaderCollision : resHeaderCollisions) {
                    List<Object> thisPath = Arrays.asList(index, "->", resHeaderCollision);
                    String thisDocumentJson = documentNamesToJson.get(documentName);
                    Object thisLocation = getPathDocumentCoordinatesPseudoJson(thisPath, thisDocumentJson);
                    parseFailures.add(new SchemaParseFailure(
                            otherDocumentName,
                            Arrays.asList(otherIndex, "->", resHeaderCollision),
                            "PathCollision",
                            Map.of("document", documentName, "path", thisPath, "location", thisLocation)));
                }
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new UApiSchemaParseError(parseFailures, documentNamesToJson);
        }
    }
}

File: tool/uapicodegen/uapicodegen/uapi/internal/schema/CatchHeaderCollisions.py
from typing import Dict, List, Set, cast


def catch_header_collisions(
    uapi_schema_name_to_pseudo_json: Dict[str, List[object]],
    header_keys: Set[str],
    keys_to_index: Dict[str, int],
    schema_keys_to_document_names: Dict[str, str],
    document_names_to_json: Dict[str, str],
) -> None:
    from ...UApiSchemaParseError import UApiSchemaParseError
    from ...internal.schema.SchemaParseFailure import SchemaParseFailure
    from ...internal.schema.GetPathDocumentCoordinatesPseudoJson import get_path_document_coordinates_pseudo_json

    parse_failures = []

    header_keys_list = list(header_keys)
    header_keys_list.sort(key=lambda k: (
        schema_keys_to_document_names[k], keys_to_index[k]))

    for i in range(len(header_keys_list)):
        for j in range(i + 1, len(header_keys_list)):
            def_key = header_keys_list[i]
            other_def_key = header_keys_list[j]

            index = keys_to_index[def_key]
            other_index = keys_to_index[other_def_key]

            document_name = schema_keys_to_document_names[def_key]
            other_document_name = schema_keys_to_document_names[other_def_key]

            uapi_schema_pseudo_json = uapi_schema_name_to_pseudo_json[document_name]
            other_uapi_schema_pseudo_json = uapi_schema_name_to_pseudo_json[other_document_name]

            def_ = cast(dict[str, object], uapi_schema_pseudo_json[index])
            other_def = cast(dict[str, object],
                             other_uapi_schema_pseudo_json[other_index])

            header_def = cast(dict[str, object], def_[def_key])
            other_header_def = cast(
                dict[str, object], other_def[other_def_key])

            header_collisions = [
                k for k in header_def if k in other_header_def]
            for header_collision in header_collisions:
                this_path = [index, def_key, header_collision]
                this_document_json = document_names_to_json[document_name]
                this_location = get_path_document_coordinates_pseudo_json(
                    this_path, this_document_json)
                parse_failures.append(
                    SchemaParseFailure(
                        other_document_name,
                        [other_index, other_def_key, header_collision],
                        'PathCollision',
                        {'document': document_name, 'path': this_path,
                            'location': this_location},
                    )
                )

            res_header_def = cast(dict[str, object], def_['->'])
            other_res_header_def = cast(dict[str, object], other_def['->'])

            res_header_collisions = [
                k for k in res_header_def if k in other_res_header_def]
            for res_header_collision in res_header_collisions:
                this_path = [index, '->', res_header_collision]
                this_document_json = document_names_to_json[document_name]
                this_location = get_path_document_coordinates_pseudo_json(
                    this_path, this_document_json)
                parse_failures.append(
                    SchemaParseFailure(
                        other_document_name,
                        [other_index, '->', res_header_collision],
                        'PathCollision',
                        {'document': document_name, 'path': this_path,
                            'location': this_location},
                    )
                )

    if parse_failures:
        raise UApiSchemaParseError(parse_failures, document_names_to_json)
